src/transpile/fusion.hpp

```diff
@@ -32,6 +32,645 @@ using oplist_t = std::vector<op_t>;
 using opset_t = Operations::OpSet;
 using reg_t = std::vector<uint_t>;
 
+class FusionMethod {
+public:
+  // Return name of method
+  virtual std::string name() = 0;
+
+  virtual bool support_diagonal() const = 0;
+
+  // Aggregate a subcircuit of operations into a single operation
+  virtual op_t generate_operation(std::vector<op_t>& fusioned_ops, bool diagonal = false) const {
+    std::set<uint_t> fusioned_qubits;
+    for (auto & op: fusioned_ops)
+      fusioned_qubits.insert(op.qubits.begin(), op.qubits.end());
+
+    reg_t remapped2orig(fusioned_qubits.begin(), fusioned_qubits.end());
+    std::unordered_map<uint_t, uint_t> orig2remapped;
+    reg_t arg_qubits;
+    arg_qubits.assign(fusioned_qubits.size(), 0);
+    for (size_t i = 0; i < remapped2orig.size(); i++) {
+      orig2remapped[remapped2orig[i]] = i;
+      arg_qubits[i] = i;
+    }
+
+    // Remap qubits
+    for (auto & op: fusioned_ops)
+      for (size_t i = 0; i < op.qubits.size(); i++)
+        op.qubits[i] = orig2remapped[op.qubits[i]];
+
+    auto fusioned_op = generate_operation_internal(fusioned_ops, arg_qubits);
+
+    // Revert qubits
+    for (size_t i = 0; i < fusioned_op.qubits.size(); i++)
+      fusioned_op.qubits[i] = remapped2orig[fusioned_op.qubits[i]];
+
+    if (diagonal) {
+      std::vector<complex_t> vec;
+      vec.assign((1UL << fusioned_op.qubits.size()), 0);
+      for (size_t i = 0; i < vec.size(); ++i)
+        vec[i] = fusioned_op.mats[0](i, i);
+      fusioned_op = Operations::make_diagonal(fusioned_op.qubits, vec, std::string("fusion"));
```

Comment:
If you add the move semantics to `make_diagonal` this should be 
```suggestion
      fusioned_op = Operations::make_diagonal(fusioned_op.qubits, std::move(vec), std::string("fusion"));
```